server/server_test.go

```diff
@@ -0,0 +1,445 @@
+package server
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"testing"
+
+	"github.com/google/cel-spec/proto/checked/v1/checked"
+	"github.com/google/cel-spec/proto/v1/cel_service"
+	"github.com/google/cel-spec/proto/v1/eval"
+	"github.com/google/cel-spec/proto/v1/syntax"
+	"github.com/google/cel-spec/proto/v1/value"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+type serverTest struct {
+	cmd    *exec.Cmd
+	conn   *grpc.ClientConn
+	client cel_service.CelServiceClient
+}
+
+var (
+	globals = serverTest{}
+)
+
+func TestMain(m *testing.M) {
+	// Use a helper function to ensure we run shutdown()
+	// before calling os.Exit()
+	os.Exit(mainHelper(m))
+}
+
+func mainHelper(m *testing.M) int {
+	err := setup()
+	defer shutdown()
+	if err != nil {
+		// testing.M doesn't have a logging method.  hmm...
+		log.Fatal(err)
+		return 1
+	}
+	return m.Run()
+}
+
+func setup() error {
+	globals.cmd = exec.Command("cel_server")
+
+	out, err := globals.cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	globals.cmd.Stderr = os.Stderr // share our error stream
+
+	log.Println("Starting server")
+	err = globals.cmd.Start()
+	if err != nil {
+		return err
+	}
+
+	log.Println("Getting server's address")
+	var addr string
+	_, err = fmt.Fscanf(out, "Listening on %s\n", &addr)
+	out.Close()
+	if err != nil {
+		return err
+	}
+
+	log.Println("Connecting to ", addr)
+	conn, err := grpc.Dial(addr, grpc.WithInsecure())
+	if err != nil {
+		return err
+	}
+	globals.conn = conn
+
+	log.Println("Creating service client")
+	globals.client = cel_service.NewCelServiceClient(conn)
+	return nil
+}
+
+func shutdown() {
+	if globals.conn != nil {
+		globals.conn.Close()
+		globals.conn = nil
+	}
+	if globals.cmd != nil {
+		globals.cmd.Process.Kill()
+		globals.cmd.Wait()
+		globals.cmd = nil
+	}
+}
+
+var (
```

Comment:
I've tried to create a series of ASTs for testing purposes in the `test/expr.go` file. Consider adding this expected value there? Also please use the helpers in that file to build the AST nodes since it makes it a bit easier to read.

Reply:
Those helpers make things much nicer, don't they? :-)  After using the helper, this giant parsed expression is pretty trivial and not worthy of reuse outside this module, IMHO.